lib/salus/scanners/osv/maven_osv.rb

```diff
@@ -0,0 +1,152 @@
+      elsif introduced.present?
+        introduced_version = SemVersion.new(introduced)
+        vulnerable_flag = true if version_found >= introduced_version
+      end
+
+      vulnerable_flag
+    end
+
+    # Compare vulnerabilities found with dependencies found
+    # and return vulnerable dependencies
+    def match_vulnerable_dependencies(dependencies)
+      results = []
+      dependencies.each do |dependency|
+        lib = "#{dependency.fetch('group_id', '')}:#{dependency.fetch('artifact_id', '')}"
+        unless dependency['version'].nil?
+          version = dependency['version']
+          # If version is of the format '${deps.version}', log it.
+          if version.match(/\${(.*)\.version}/)
+            bugsnag_notify("MavenOSV: Found #{lib}:#{version} with incompatible format.")
+            next
+          end
+
+          package_matches = @osv_vulnerabilities.select do |v|
+            v.dig("package", "name") == lib
+          end
+          package_matches.each do |m|
+            m["ranges"].each do |version_ranges|
+              introduced = fixed = ""
+              if version_ranges["events"].length == 1
+                if version_ranges["events"][0].key?("introduced")
+                  introduced = version_ranges["events"][0]["introduced"]
+                end
+              elsif version_ranges["events"].length == 2
+                if version_ranges["events"][0].key?("introduced") &&
+                    version_ranges["events"][1].key?("fixed")
+                  introduced = version_ranges["events"][0]["introduced"]
+                  fixed = version_ranges["events"][1]["fixed"]
+                end
+              end
+
+              if version_ranges["type"] == "SEMVER" || version_ranges["type"] == "ECOSYSTEM"
+                if version_matching(version, introduced, fixed)
+                  results.append({
+                                   "Package": m.dig("package", "name"),
+                              "Vulnerable Version": introduced,
+                              "Version Detected": version,
+                              "Patched Version": fixed,
+                              "ID": m.fetch("aliases", [m.fetch("id", [])])[0],
+                              "Database": m.fetch("database"),
+                              "Summary": m.fetch("summary", m.dig("details")).strip,
+                              "References": m.fetch("references", []).collect do |p|
+                                              p["url"]
+                                            end.join(", "),
+                              "Source":  m.dig("database_specific", "url") || DEFAULT_SOURCE,
+                              "Severity": m.dig("database_specific", "severity") || DEFAULT_SEVERITY
+                                 })
+                end
+              end
+            end
+          end
+        end
+      end
+      results
+    end
+
+    # Find dependencies from the project
+    def find_dependencies
```

Comment:
This block is a copy of https://github.com/coinbase/salus/blob/master/lib/salus/scanners/report_pom_xml.rb#L10-L25

If we created a  MavenParser we could DRY the code up.  Really the bin/parse_pom_xml logic could eventually be moved into that class.  If it memoized (we would need a mutex to keep our concurrent code happy) the result we could also avoid parsing the same file 2x.  

Thoughts?

Reply:
Yup, it makes sense if we moved it to its own class, currently just followed the existing structure but yes makes sense. This will become even more apparent when we do a gradle parser because it takes a long time so having one class return data to both CVE Scanners and Reporting Scanners would be nice.